### PR TITLE
Add database type adjustments to server side generated temporal values

### PR DESCRIPTION
**Pull request by lbrdnk**
Closes #46924

This PR fixes casting of timestamps to server side relative datetime logic on Snowflake. Datetime are casted to type of corresponding field.

### CODE REVIEW

**Review by bshepherdson — COMMENTED** [review]
While I agree that needing to do this sucks, I don't think it's an abuse of dynamic vars. I'm not aware of a better way to achieve this.

However, I'm concerned that needing to know about this `binding` and do it in the right places is a complexity I'd be glad to avoid. It would be great to do something like a Methodical "around" wrapper than did the binding automatically, in one place, for all expressions. But the generic `->honeysql` is too broad, so I'm not sure it's practical.

**Comment by metamben** [comment]
> While I agree that needing to do this sucks, I don't think it's an abuse of dynamic vars. I'm not aware of a better way to achieve this.

I think the usual approach is to annotate the value to be transformed with the information from the context. Could we do that here?

**Comment by lbrdnk** [comment]
> While I agree that needing to do this sucks, I don't think it's an abuse of dynamic vars. I'm not aware of a better way to achieve this.
> 
> However, I'm concerned that needing to know about this `binding` and do it in the right places is a complexity I'd be glad to avoid. It would be great to do something like a Methodical "around" wrapper than did the binding automatically, in one place, for all expressions. But the generic `->honeysql` is too broad, so I'm not sure it's practical.

Thank you for the feedback. What you mention would be ideal. In this case however it may be problematic. The `:=` implementation first transforms the field prior to honeysql call. I believe the wrapper would need to be passed that transforming function.

The places where the binding should occur are described in the vars docstring.

File: modules/drivers/snowflake/src/metabase/driver/snowflake.clj [hunk 1]
```diff
@@ -670,7 +672,10 @@
   (sql-jdbc.execute/do-with-connection-with-options
    driver database nil
    (fn [^java.sql.Connection conn]
-     (with-open [stmt (.prepareStatement conn "show parameters like 'TIMEZONE' in user;")
+     (with-open [stmt (.prepareStatement conn (str "show parameters like 'TIMEZONE' in user"
+                                                   (when-let [user (get-in database [:details :user])]
+                                                     (str " \"" (str/replace user #"\s" "") "\""))
```
Comment (metamben) [inline modules/drivers/snowflake/src/metabase/driver/snowflake.clj:677]: This reminds me of https://xkcd.com/327/

I don't know what we want to achieve here, but it does look a bit dangerous.

File: modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj [hunk 2]
```diff
@@ -811,3 +812,89 @@
...
+
+;;;;
+;;;; GOOD DATETIMES IN BELIZE
+;;;; Testing Snowflake timestamp types in relative date time filter.
+;;;; (In Belize they know no DST anymore.)
+;;;;
+
+(defn- rows-for-good-datetimes-in-belize
+  []
+  (let [belize-offset (t/zone-offset "-06:00")
```
Comment (metamben) [inline modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj:824]: Extract to top level?

File: modules/drivers/snowflake/test/metabase/test/data/snowflake.clj [hunk 3]
```diff
@@ -152,15 +152,15 @@
...
+(defn- set-user-timezone!
+  [user timezone]
   (sql-jdbc.execute/do-with-connection-with-options
    :snowflake
    (no-db-connection-spec)
    {:write? true}
    (fn [^java.sql.Connection conn]
      (with-open [stmt (.createStatement conn)]
-       (.execute stmt (format "ALTER USER SET TIMEZONE = '%s';" timezone))))))
+       (.execute stmt (format "ALTER USER \"%s\" SET TIMEZONE = '%s';" (str/replace user #" " "") timezone))))))
```
Comment (metamben) [inline modules/drivers/snowflake/test/metabase/test/data/snowflake.clj:163]: Should this be consistent with the other use name manipulation?

**Review by metamben — APPROVED** [review]
Did it turn out there was no such thing as a "current" user?

File: modules/drivers/snowflake/src/metabase/driver/snowflake.clj (hunk 1 above)
Comment (lbrdnk) [reply modules/drivers/snowflake/src/metabase/driver/snowflake.clj:677]: Github should have a badge for receiving a review comment with link to that xkcd. Reading your comment I've realized that change is probably redundant. When I confirm that, I'll revert this and other related bit. Thanks.

**Pull request merged**